### PR TITLE
Defer creation of Location objects to the point they are actually needed.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Most commonly, we are creating locations so we can report diagnostics at that place in the code.  However, teh common case is no diagnostics, so these allocs are wasteful.   Addresses many top codepaths that run through:

image

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Declarations/SingleTypeDeclaration.cs [hunk 1]
```diff
@@ -245,7 +245,7 @@ public bool Equals(TypeDeclarationIdentity other)
                     return false;
                 }
 
-                if ((object)thisDecl.Location.SourceTree != otherDecl.Location.SourceTree
+                if ((object)thisDecl.SyntaxReference.SyntaxTree != otherDecl.SyntaxReference.SyntaxTree
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Declarations/SingleTypeDeclaration.cs:248]: the Location was created from teh SyntaxReference, so this just avoids the unnecessary middle alloc.

File: src/Compilers/CSharp/Portable/Errors/MessageID.cs [hunk 2]
```diff
@@ -317,13 +317,12 @@ public static LocalizableErrorArgument Localize(this MessageID id)
         internal static bool CheckFeatureAvailability(
             this MessageID feature,
             BindingDiagnosticBag diagnostics,
-            SyntaxNode syntax,
+            SyntaxNodeOrToken syntax,
             Location? location = null)
         {
-            var diag = GetFeatureAvailabilityDiagnosticInfo(feature, (CSharpParseOptions)syntax.SyntaxTree.Options);
-            if (diag is object)
+            if (GetFeatureAvailabilityDiagnosticInfo(feature, (CSharpParseOptions)syntax.SyntaxTree!.Options) is { } diagInfo)
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Errors/MessageID.cs:323]: updated to match the method two below.

**Comment by CyrusNajmabadi** [comment]
@RikkiGibson @333fred ptal.

**Review by 333fred — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Binder/Binder.cs [hunk 3]
```diff
@@ -748,12 +748,12 @@ internal static void ReportDiagnosticsIfUnmanagedCallersOnly(BindingDiagnosticBa
                 // In the former case, we use a lazy diagnostic that may end up being ignored later, to avoid causing a
                 // binding cycle.
                 diagnostics.Add(unmanagedCallersOnlyAttributeData == UnmanagedCallersOnlyAttributeData.Uninitialized
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Binder/Binder.cs:750]:
```suggestion
                var location = syntax.GetLocation();
                Debug.Assert(location != null);
                diagnostics.Add(unmanagedCallersOnlyAttributeData == UnmanagedCallersOnlyAttributeData.Uninitialized
```

File: src/Compilers/CSharp/Portable/Binder/Binder.cs [hunk 4]
```diff
@@ -739,21 +739,23 @@ internal static void ReportDiagnosticsIfObsoleteInternal(BindingDiagnosticBag di
...
+                Debug.Assert(location != null);
                 diagnostics.Add(unmanagedCallersOnlyAttributeData == UnmanagedCallersOnlyAttributeData.Uninitialized
-                                    ? (DiagnosticInfo)new LazyUnmanagedCallersOnlyMethodCalledDiagnosticInfo(symbol, isDelegateConversion)
+                                    ? new LazyUnmanagedCallersOnlyMethodCalledDiagnosticInfo(symbol, isDelegateConversion)
                                     : new CSDiagnosticInfo(isDelegateConversion
                                                                ? ErrorCode.ERR_UnmanagedCallersOnlyMethodsCannotBeConvertedToDelegate
                                                                : ErrorCode.ERR_UnmanagedCallersOnlyMethodsCannotBeCalledDirectly,
                                                            symbol),
-                                location);
+                                syntax.GetLocation()!);
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Binder/Binder.cs:758]:
```suggestion
                                location);
```

File: src/Compilers/CSharp/Portable/Binder/Binder.cs [hunk 5]
```diff
@@ -739,16 +739,18 @@ internal static void ReportDiagnosticsIfObsoleteInternal(BindingDiagnosticBag di
...
-        internal static void ReportDiagnosticsIfUnmanagedCallersOnly(BindingDiagnosticBag diagnostics, MethodSymbol symbol, Location location, bool isDelegateConversion)
+        internal static void ReportDiagnosticsIfUnmanagedCallersOnly(BindingDiagnosticBag diagnostics, MethodSymbol symbol, SyntaxNodeOrToken syntax, bool isDelegateConversion)
         {
             var unmanagedCallersOnlyAttributeData = symbol.GetUnmanagedCallersOnlyAttributeData(forceComplete: false);
             if (unmanagedCallersOnlyAttributeData != null)
             {
                 // Either we haven't yet bound the attributes of this method, or there is an UnmanagedCallersOnly present.
                 // In the former case, we use a lazy diagnostic that may end up being ignored later, to avoid causing a
                 // binding cycle.
+                var location = syntax.GetLocation();
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Binder/Binder.cs:750]: sorry, i'm not sure what you're asking for here.  Can you clarify?  what do you mean by 'void diagnostic'?

[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Binder/Binder.cs:750]: Clarified offline.

[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Binder/Binder.cs]: ok.  i moved it into the assert.  but i don't think it helps.  can you take another look fred :)

**Review by 333fred — APPROVED** [review]

**Comment by CyrusNajmabadi** [comment]
@jaredpar ptal :)

**Comment by CyrusNajmabadi** [comment]
@RikkiGibson ptal :)

**Comment by CyrusNajmabadi** [comment]
@RikkiGibson @jaredpar ptal :)

**Review by jaredpar — APPROVED** [review]

**Pull request merged**